### ISSUE TITLE
Add manual property syncing to MultiplayerSynchronizer

### ISSUE THREAD

**Issue by DigitallyTailored** [godotengine/godot-proposals#9406]
### Describe the project you are working on

An open world game that primarily runs from a server that clients connect to. Almost all nodes can be updated/moved/rotated by clients as the world is highly interactive.

### Describe the problem or limitation you are having in your project

The options for syncing properties of MultiplayerSynchronizer are limited. We can either set them to `never` update which is useful for when a node is spawned - to `always` update which syncs properties across the network at given intervals regardless of changes but potentially uses a lot of bandwidth, or on `watch` where the engine watches these properties for changes. `Watch` would be ideal but when thousands of entities are being watched this causes a noticeable drop in performance. Additionally, performance seems to halve for each new client added - likely as internally it's storing the state of all properties per connected peer and constantly comparing them.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a manual `sync_property` method that takes the property reference and broadcasts it to all visibile peers. This allows for implementing custom sync logic per property and per node, making sure we only sync properties when we know that they have changed without relying on godot to `watch` everything per client.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Lets say I have an RTS with a few hundred units. Most of the time I don't want to broadcast from every unit constantly and monitoring all of their properties for changes will causes noticeable slowdown. Instead, when a state changes for a unit or group of units, say I set a new target destination for them, as part of this I could also call `sync_property` to bypass the overhead of having those properties watched (when I already know that the property is about to change):

```
for unit in units:
    target_vector = <some new position>
    unit.multiplayer_synchronizer.sync_property(target_vector)
```

Another scenario - which is one I'm currently facing - is if I have a terrain node/chunk that contains a small heightmap. I only want to sync the heightmap on spawn and if the terrain has been deformed. This is not a key part of my game so it seems wasteful to either constantly broadcast that heightmap data at inervals or to add the overhead of watching the heightmap data property. Instead, when the terrain is deformed and I perform that action on the server, it makes sense that I can manually broadcast that this change has happened immediately.

Finally, I may have lots of environmemt scenes such as walls. If I destroy one I would want to set the state to `destroyed` and broadcast that immediately, but I don't want to `watch` hundreds of walls and I only want to send an update once:
```
wall.multiplayer_synchronizer.sync_property("wall_status", "destroyed")
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Possibly by toggling the current visibility state but it's not clear if that would cause an immediate sync before it's toggled off again.

We could use an Rpc call to set the property and broadcast and handle the update from there but as we already have the property added to the multiplayerSynchronizer for the reason of syncing it across the network I think it should be possible from there.

### Is there a reason why this should be core and not an add-on in the asset library?

It seems like a very small addition to the current multiplayerSynchronizer functionality that allows for much greater control.

**Comment by DigitallyTailored** [comment]
I don't think there is an issue using Rpc beyond it being a separate way to do what appears to be the _almost_ the same thing as `on change` already does. The biggest benefit over using an Rpc and having a method like this is perhaps consolidation of configuration - properties are already configured and managed from one place and we can tap into the multiplayerSynchronizer for the manual syncs like we already do for specifying update intervals and sync types per property without having to additionally set up Rpc calls.

> ... so I think having properties that only change on manual request would be using a very unnecessary tool to do so

I appreciate that  - we do however already have an `on change` setting which does the same thing but with the overhead of looking out for changes, which as shown above is costly. I think manually telling the multiplayerSynchronizer to sync a property for a node fits somewhere between the existing `watch` and `on change` options.

**Comment by DigitallyTailored** [comment]
Sorry - I should have been clearer. By 'doing the same thing' I mean that it sends out an update for that property as a consequence of it detecting a property changing. I think specifically I'm suggesting being able to manually tell it to broadcast that property at any given time - which it appears to do already at set intervals with the `always` setting for instance.

OR if the workflow is completely different behind the scenes - I have no idea - then perhaps the `on change` config option is not meant to have this much overhead, in which case a manual sync method would not be needed at all.

**Comment by Withaust** [comment]
I accidentally created identical issue above while trying out new netcode for the first time. I completely agree with @DigitallyTailored that having manual polling belongs in `MultiplayerSynchronizer`, since visibility is also important in lowering bandwith usage. If you were going to implement this with RPCs, you would have to recreate this entire system from scratch on your own, which is the complete opposite of what you want out of the networking solution.

**Comment by mackerelshark** [comment]
Did anyone ever figure out a work-around for this?  It doesn't look like it's going to get put into the project anytime soon (totally understand the devs have a huge workload) but it would be nice to be able to use the easy functionality of the MultiplayerSynchronizer node, with the efficiency of a manual trigger, instead of the game devoting resources to monitoring the properties for changes.

**Comment by DigitallyTailored** [comment]
@mackerelshark my solution in the meantime is, ontop of a MultiplayerSynchronizer, wrapping all my scenes in a custom sync node that has an RPC function for manually setting and broadcasting properties. I went this way as a lot of my scenes don't even have scripts but can be picked up and moved etc and it allows me to keep scripts I _do_ add very simple with little consideration for networking. I can also add interpolation in here too which automatically gets applied to all scenes when used.

My Sync scene:

![Image](https://github.com/user-attachments/assets/339d8676-0cb1-4520-bc0f-a9478fccbd2d)

![Image](https://github.com/user-attachments/assets/c3355a6e-09d5-4660-9038-14d54d4f37d6)

(These are all populated in a script as part of my procgen process)

In gameplay my scene tree looks like this, you can see in the background I'm manually loading in the scene under this:

![Image](https://github.com/user-attachments/assets/6455e466-876b-4925-82de-2177d45569c2)

My scene gets added as a child of this node. I'm basically just using the MultiplayerSynchronizer for visibility - which it's great at - plus having automatically shared scene paths bewtween clients is quite handy.

I still think manual synching/dirtying would be a great addition. As it stands I'd probably try and replace MultiplayerSynchronizer with straight up RPC if I have to revisit networking, but this is good enough for now.

**Comment by Withaust** [comment]
@DigitallyTailored Do you still have your MRP that makes it apparent with its FPS that there is a performance issue with the "On Changed" logic? I am planning to tackle this later in a few days and I am trying to collect as many MRPs as possible to test my PR on.

**Comment by DigitallyTailored** [comment]
Hi @Withaust - that sounds great! I think the one attached above is all good https://github.com/godotengine/godot-proposals/files/14805358/multiplayer_perf.zip

https://github.com/godotengine/godot-proposals/issues/9406#issuecomment-2027341089

Best of luck!

**Comment by Withaust** [comment]
> Hi [@Withaust](https://github.com/Withaust) - that sounds great! I think the one attached above is all good https://github.com/godotengine/godot-proposals/files/14805358/multiplayer_perf.zip
> 
> [#9406 (comment)](https://github.com/godotengine/godot-proposals/issues/9406#issuecomment-2027341089)
> 
> Best of luck!

Yeah, nevermind, I tried but unfortunately solutions to this issue requires some additional design decisions:

How to properly implement dirtying of the variables from the API point of view?
Should we add additional synchronization intervals?
Should manual dirtying even be considered or should we try fixing underlying issue with "On Changed" hashing?

I could start working on a PR, but I dont want to spend time implementing something for it to be rejected... This have to be talked through with some of the people that are responsible for maintaining networking first

**Comment by Ryan-000** [comment]
In my project, we’re currently using around 950 synchronizers. Even though most of them are using delta-sync, I’ve still noticed many performance issues. According to my C# profiler, around 35% of my managed time is spent handling synchronizer delta-property overhead.

And with a native profiler, `SceneReplicationInterface::on_network_process()` accounts for about 60% of the time inside `SceneTree::process`

![Image](https://github.com/user-attachments/assets/83c99446-61f0-4887-9f27-75b4374f7b1e)

Another issue is that that synchronizers are always updated within `SceneReplicationInterface::on_network_process()` which does not respect their process mode. This means that the only reliable way to disable them is to remove them from the SceneTree, which is not possible or feasible in many scenarios.

Looking into the native implementation, there is still a lot of room for optimizations to reduce double work, minimize dynamic allocations, and add caching. Moreover, these operations are repeated per peer, which has poor scalability, as @DigitallyTailored pointed out.

**Comment by Withaust** [comment]
I think I figured out a temporary solution to this problem. 

You can derive from MultiplayerSynchronizer and do the following:

1. Set your target property to be replicated "On Spawn" and "On Changed".

2. Set timer interval for "On Changed" replication to some really big number 

3. When you want to dirty and send a current state immediately - you do so with calling a custom dirty() method in class deriving of MultiplayerSynchronizer that simply puts the "On Changed" replication timer to be 0.0 (immediate)

4. In the next "_process" call you can set back the replication interval to some large number so that it wont do any unnecessary expensive Variant hash comparing (or wait an additional frame to make sure that this change was definitely acknowledged by the "on Changed" replication)

I know this does not fix the underlying issue, but at the very least it emulates what I had in mind with the dirtying approach
